db/migrations/20200603124229_create_time_ilk_snapshot_query.sql

```diff
@@ -0,0 +1,66 @@
+-- +goose Up
+CREATE TYPE api.time_ilk_snapshot AS
+(
+    bucket_start TIMESTAMP,
+    bucket_end TIMESTAMP,
+    bucket_interval INTERVAL,
+    ilk_identifier TEXT,
+    block_number BIGINT,
+    rate NUMERIC,
+    art NUMERIC,
+    spot NUMERIC,
+    line NUMERIC,
+    dust NUMERIC,
+    chop NUMERIC,
+    lump NUMERIC,
+    flip TEXT,
+    rho NUMERIC,
+    duty NUMERIC,
+    pip TEXT,
+    mat NUMERIC,
+    created TIMESTAMP,
+    updated TIMESTAMP
+);
+
+CREATE FUNCTION api.time_ilk_snapshots(ilk_identifier TEXT, range_start TIMESTAMP, range_end TIMESTAMP, bucket_interval INTERVAL DEFAULT '1 day'::INTERVAL)
+    RETURNS SETOF api.time_ilk_snapshot AS
+$$
+WITH buckets AS (SELECT generate_series(range_start, range_end - bucket_interval, bucket_interval) AS bucket_start)
```

Comment:
Curious about a couple things - 
1) why does the series end at `range_end - bucket_interval` instead of just `range_end`?
2) why do we select `AS bucket_start`?

Reply:
1. `generate_series()` uses inclusive start and end times, and with my use case I've been using interval aligned parameters which are slightly easier to iterate over when the end time is exclusive. This was a stylistic choice. Thinking on it now, it's probably more intuitive to keep the range_end as inclusive as you suggest. I'll make that change.
2. This is to name the column `bucket_start` instead of the auto-generated `generate_series` as I reference in `buckets.bucket_start`. Would you prefer another style?

Reply:
Update on 1:

I just remembered my original thought process. To me I wanted to include all full interval periods between the start and end date, and did not want to include any last interval that is not fully contained within the date range. For example, if I did not remove a `bucket_interval` from the last date then `time_ilk_snapshots('ETH-A', '2020-01-01 00:00:00', '2020-01-31 00:00:00', '1 day'::INTERVAL)` would return with the last `buckets` row being `bucket_start = 2020-01-31` and `bucket_end = 2020-02-01` which feels unintuitive to me because the end date was specified as midnight at the start of the Jan 31st and not midnight at the start of Feb 1st.